File: src/MaksIT.CertsUI/Models/APIKeys/PatchApiKeyRequest.cs
using MaksIT.Core.Abstractions.Webapi;

namespace MaksIT.CertsUI.Models.APIKeys;

public class PatchApiKeyRequest : PatchRequestModelBase {

  public bool? IsGlobalAdmin { get; set; }
  public List<PatchApiKeyEntityScopeRequest>? EntityScopes { get; set; }

  public string? Description { get; set; }

  public DateTime? ExpiresAt { get; set; }
}

File: src/MaksIT.WebUI/src/models/apiKey/PatchApiKeyRequest.ts
import { array, boolean, object, string, ZodType } from 'zod'
import { PatchRequestModelBase, PatchRequestModelBaseSchema } from '@maks-it.com/webui-contracts'
import { withPatchGlobalAdminRefine } from '../engine/globalAdminZod'
import { PatchApiKeyEntityScopeRequest, PatchApiKeyEntityScopeRequestSchema } from './PatchApiKeyEntityScopeRequest'

export interface PatchApiKeyRequest extends PatchRequestModelBase {
  description?: string | null
  expiresAt?: string | null
  isGlobalAdmin?: boolean | null
  entityScopes?: PatchApiKeyEntityScopeRequest[] | null
}

export const PatchApiKeyRequestSchema: ZodType<PatchApiKeyRequest> = PatchRequestModelBaseSchema.and(
  object({
    description: string().optional().nullable(),
    expiresAt: string().optional().nullable(),
    isGlobalAdmin: boolean().optional().nullable(),
    entityScopes: array(PatchApiKeyEntityScopeRequestSchema).optional().nullable(),
  })
)

export const createPatchApiKeyRequestSchema = (actorIsGlobalAdmin: boolean) =>
  withPatchGlobalAdminRefine(PatchApiKeyRequestSchema, actorIsGlobalAdmin)
